src/main/java/com/netgrif/application/engine/workflow/web/AbstractTaskController.java

```diff
@@ -199,16 +199,17 @@ public CountResponse count(SingleElasticTaskSearchRequestAsList query, MergeFilt
     public EntityModel<EventOutcomeWithMessage> setData(String taskId, ObjectNode dataBody) {
         Map<String,SetDataEventOutcome> outcomes = new HashMap<>();
         dataBody.fields().forEachRemaining(it -> outcomes.put(it.getKey(), dataService.setData(it.getKey(), it.getValue().deepCopy())));
-        SetDataEventOutcome mainOutcome = outcomes.get(taskId);
-        outcomes.remove(taskId);
-        mainOutcome.addOutcomes(new ArrayList<>(outcomes.values()));
+        SetDataEventOutcome mainOutcome = taskService.getMainOutcome(outcomes, taskId);
         return EventOutcomeWithMessageResource.successMessage("Data field values have been sucessfully set",
                 LocalisedEventOutcomeFactory.from(mainOutcome, LocaleContextHolder.getLocale()));
     }
 
-    public EntityModel<EventOutcomeWithMessage> saveFile(String taskId, String fieldId, MultipartFile multipartFile) {
+    public EntityModel<EventOutcomeWithMessage> saveFile(String taskId, String fieldId, MultipartFile multipartFile, ObjectNode dataBody) {
```

Comment:
Do not use generic `ObjectNode`. Create request body class to hold received data.

Reply:
changed to Map<String, String>